Switch to sirbrillig/phpcs-variable-analysis

GaryJones opened the pull request:
Introduces a new dependency of `sirbrillig/phpcs-variable-analysis` which is a more maintained version of the existing `VariableAnalysis` sniff in VIPCS.

The new sniff is added to the `WordPressVIPMinimum` ruleset. `WordPress-VIP-Go` ruleset inherits this, and maintains its dropping of the severity for `UnusedVariable` and `UndefinedVariable` violations.

The ruleset tests have set the previously undefined variables (or changed some variable names), so that violations on each line are still specific to the thing being tested, and doesn't include extra violations due to the use of throwaway variables.

One of the reasons for changing was the previously false positive with respect to the use of `$this` inside closures inside class methods. A ruleset test has been added to cover this, and it correctly does not get flagged in the ruleset test.

The old `VariableAnalysis` has been marked as deprecated, until it can be removed in the next major release of VIPCS. This will stop any breakage of setups which reference `WordPressVIPMinimum.Variables.VariableAnalysis` in a custom ruleset. If called, it throws a warning, and then passes control to the new VariableAnalysis sniff. The deprecation notice is excluded if the sniff is not explicitly included, and the duplicate violation messages are also silenced.

The previously used `VariableAnalaysisHelper.php` has been removed completely, as no-one should have been referencing that directly.

Fixes #449.

rebeccahum commented on WordPressVIPMinimum/ruleset.xml:
```diff
@@ -157,10 +157,35 @@
+		<severity>0</severity>
+	</rule>
+
+	<!-- Prevent duplicate messages from deprecated sniff. -->
```
I'm a bit confused, how come we want to silence these below warnings if they are being removed in the VariableAnalysisSniff?

GaryJones replied:
This is how deprecations are done in WPCS. If I comment these out, I do get duplicate messages, as PHPCS has registered the same violations under two different sniff classes (albeit that both ultimately use the same logic from the new package class).

GaryJones commented:
> Is there a way to create an "alias", so this won't break the `// phpcs:ignore WordPressVIPMinimum.Variables.VariableAnalysis.UndefinedVariable` comments?

Not as far as I know. The deprecation handles references in custom rulesets, but not ignore / disable references.

GaryJones commented:
Noting that VA 2.6.3+ requires PHP 5.6, so a composer install may need to ignore platform requirements to be able to test PHP 5.4 and PHP 5.5.

GaryJones commented:
Refactored and fixed up tests and updated to use VA `^2.8.2`. Currently failing on PHP 5.4 and PHP 5.5 Travis jobs.

I tested ignoring platform requirements so that VA could be installed on PHP 5.4 and 5.5, but a) it fails due to the PHP 5.6+ code in VA, and b) ignoring platform requirements make no sense when this is a non-dev dependency.

I suspect my [comment](https://github.com/sirbrillig/phpcs-variable-analysis/pull/86#issuecomment-657076732) isn't going to result in VA moving back to PHP 5.4 due to the code in it, so we'll have to bump the minimum PHP for VIPCS back up (having only [just dropped it](https://github.com/Automattic/VIP-Coding-Standards/pull/491)) to PHP 5.6.

GaryJones commented:
Before merging, we should check the changes here against the proposals made in https://github.com/Automattic/VIP-Coding-Standards/issues/449, or at least open a follow-up PR which the extra changes, within the same release.

sirbrillig commented:
> I suspect my comment isn't going to result in VA moving back to PHP 5.4 due to the code in it, so we'll have to bump the minimum PHP for VIPCS back up (having only just dropped it) to PHP 5.6.

Surprise! 😁  https://github.com/sirbrillig/phpcs-variable-analysis/pull/191

I'll still need to backport it to the 2.x branch though. Coming soon...

sirbrillig commented:
https://github.com/sirbrillig/phpcs-variable-analysis/releases/tag/v2.8.3 👍

GaryJones commented:
@sirbrillig You are a wonderful human - thank you!
